Mathlib/Topology/CWComplex.lean

```diff
@@ -0,0 +1,164 @@
+notation "𝔻 "n => disk n
+
+/-- The inclusion map from the n-sphere to the (n+1)-disk -/
+def sphereInclusion (n : ℤ) : (𝕊 n) ⟶ (𝔻 n + 1) where
+  toFun := fun ⟨p, hp⟩ ↦ ⟨p, le_of_eq hp⟩
+  continuous_toFun := ⟨fun t ⟨s, hso, hst⟩ ↦ by
+    rw [isOpen_induced_iff, ← hst]
+    tauto⟩
+
+/-- The inclusion map from the disjoint union of n-spheres to the disjoint union of (n+1)-disks,
+where both of the disjoint unions are indexed by `cells` -/
+def sigmaSphereInclusion (n : ℤ) (cells : Type) :
+    TopCat.of (Σ (_ : cells), 𝕊 n) ⟶ TopCat.of (Σ (_ : cells), 𝔻 n + 1) where
+  toFun := Sigma.map id fun _ x ↦ (sphereInclusion n).toFun x
+  continuous_toFun := Continuous.sigma_map fun _ ↦ (sphereInclusion n).continuous_toFun
+
+/-- Given an attaching map for each n-sphere, we construct the attaching map for the disjoint
+union of the n-spheres. -/
+def sigmaAttachMap (X : TopCat) (n : ℤ) (cells : Type) (attach_maps : cells → C(𝕊 n, X)) :
+    TopCat.of (Σ (_ : cells), 𝕊 n) ⟶ X where
+  toFun := fun ⟨i, x⟩ ↦ attach_maps i x
+  continuous_toFun := continuous_sigma fun i ↦ (attach_maps i).continuous_toFun
+
+/-- A type witnessing that X' is obtained from X by attaching (n+1)-disks -/
+structure AttachCells (X X' : TopCat) (n : ℤ) where
+  /-- The index type over the (n+1)-disks -/
+  cells : Type
+  /-- For each (n+1)-disk, we have an attaching map from its boundary, namely an n-sphere,
+  to `X`. -/
+  attach_maps : cells → C(𝕊 n, X)
+  /-- X' is the pushout obtained from X along `sigmaAttachMap`. -/
+  iso_pushout : X' ≅ Limits.pushout
+    (sigmaSphereInclusion n cells) (sigmaAttachMap X n cells attach_maps)
+
+end CWComplex
+
+/-- A relative CW-complex contains an expanding sequence of subspaces `sk i`
+(called the `i`-skeleta) for `i ≥ -1`, where `sk (-1)` is an arbitrary topological space,
+isomorphic to `A`, and each `sk (n+1)` is obtained from `sk n` by attaching (n+1)-disks. -/
+structure RelativeCWComplex (A : TopCat) where
+  /-- Skeleta -/
+  sk : ℤ → TopCat
+  /-- A is isomorphic to the (-1)-skeleton. -/
+  iso_sk_neg_one : A ≅ sk (-1)
+  /-- The (n+1)-skeleton is obtained from the n-skeleton by attaching (n+1)-disks. -/
+  attach_cells : (n : ℤ) → CWComplex.AttachCells (sk n) (sk (n + 1)) n
+
+/-- A CW-complex is a relative CW-complex whose (-1)-skeleton is empty. -/
+abbrev CWComplex := RelativeCWComplex (TopCat.of Empty)
+
+namespace CWComplex
+
+noncomputable section Topology
+
+/-- The inclusion map from `X` to `X'`, given that `X'` is obtained from X by attaching
+(n+1)-disks -/
+def attachCellsInclusion (X X' : TopCat) (n : ℤ) (att : AttachCells X X' n) : X ⟶ X' :=
+  @Limits.pushout.inr TopCat _ _ _ X
+    (sigmaSphereInclusion n att.cells)
+    (sigmaAttachMap X n att.cells att.attach_maps) _ ≫ att.iso_pushout.inv
+
+/-- The inclusion map from the n-skeleton to the (n+1)-skeleton of a relative CW-complex -/
+def skeletaInclusion {A : TopCat} (X : RelativeCWComplex A) (n : ℤ) : X.sk n ⟶ X.sk (n + 1) :=
```

Comment:
This should be in the `RelativeCWComplex` namespace.

Reply:
Thanks. This definition has now been renamed `RelativeCWComplex.inclusion`.
https://github.com/leanprover-community/mathlib4/blob/3cfda01c25cc8ef045b74478e0ccc1f6dbb3571c/Mathlib/Topology/CWComplex.lean#L127-L129